libraries/AP_MultiHeap/AP_MultiHeap.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+  multiple heap interface, allowing for an allocator that uses
+  multiple underlying heaps to cope with multiple memory regions on
+  STM32 boards
+ */
+
+#include "AP_MultiHeap.h"
+
+#if ENABLE_HEAP
+#include <AP_Math/AP_Math.h>
+#include <stdio.h>
+
+/*
+  allow up to 10 heaps
+ */
+#ifndef MAX_HEAPS
+#define MAX_HEAPS 10
+#endif
+
+extern const AP_HAL::HAL &hal;
+
+/*
+  create heaps with a total memory size, splitting over at most
+  max_heaps
+ */
+bool MultiHeap::create(uint32_t total_size, uint8_t max_heaps, bool _allow_expansion, uint32_t _reserve_size)
+{
+    max_heaps = MIN(MAX_HEAPS, max_heaps);
+    if (heaps != nullptr) {
+        // don't allow double allocation
+        return false;
+    }
+    heaps = NEW_NOTHROW Heap[max_heaps];
+    if (heaps == nullptr) {
+        return false;
+    }
+    num_heaps = max_heaps;
+    for (uint8_t i=0; i<max_heaps; i++) {
+        uint32_t alloc_size = total_size;
+        while (alloc_size > 0) {
+            heaps[i].hp = heap_create(alloc_size);
+            if (heaps[i].hp != nullptr) {
+                total_size -= alloc_size;
+                sum_size += alloc_size;
+                break;
+            }
+            alloc_size *= 0.9;
+        }
+        if (total_size == 0) {
+            break;
+        }
+    }
+    if (total_size != 0) {
+        destroy();
+        return false;
+    }
+
+    allow_expansion = _allow_expansion;
+    reserve_size = _reserve_size;
+
+    return true;
+}
+
+// destroy heap
+void MultiHeap::destroy(void)
+{
+    if (!available()) {
+        return;
+    }
+    for (uint8_t i=0; i<num_heaps; i++) {
+        if (heaps[i].hp != nullptr) {
+            heap_destroy(heaps[i].hp);
+            heaps[i].hp = nullptr;
+        }
+    }
+    delete[] heaps;
+    heaps = nullptr;
+    num_heaps = 0;
+    sum_size = 0;
```

Comment:
```suggestion
    sum_size = 0;
    expanded_to = 0;
```

Otherwise you can't restart scripting to clear the error even if you've increased the heap size.